[tflite] export SetNumThreads to TFLite Python API

freedomtan opened the pull request:
export SetNumThreads() and add an option to set the number of threads in the `label_image.py` for TFLite. On [PYNQ-Z1](https://www.xilinx.com/products/boards-and-kits/1-hydd4z.html), a board with 2xCA9,
```
$ python tensorflow/lite/examples/python/label_image.py --num_threads 1
0.415686: 653:military uniform
0.352941: 907:Windsor tie
0.058824: 668:mortarboard
0.035294: 458:bow tie, bow-tie, bowtie
0.035294: 835:suit, suit of clothes
time:  0.691580057144

$ python tensorflow/lite/examples/python/label_image.py --num_threads 2
0.415686: 653:military uniform
0.352941: 907:Windsor tie
0.058824: 668:mortarboard
0.035294: 458:bow tie, bow-tie, bowtie
0.035294: 835:suit, suit of clothes
time:  0.412344932556
```

freedomtan committed:
[tflite] export SetNumThreads to TFLite Python API

export SetNumThreadS() and add an option to set the number of
threads in the `label_image.py` for TFLite.

jianlijianli commented:
added @miaout17 to the reviewer list since it involves interpreter change.

jdinkel88 commented:
Mine built well, but I get an object has no attribute
Traceback (most recent call last):
  File "./label_image.py", line 74, in <module>
    interpreter.set_num_threads(int(args.num_threads))
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/contrib/lite/python/interpreter.py", line 260, in set_num_threads
    return self._interpreter.SetNumThreads(i)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/contrib/lite/python/interpreter_wrapper/tensorflow_wrap_interpreter_wrapper.py", line 97, in <lambda>
    __getattr__ = lambda self, name: _swig_getattr(self, InterpreterWrapper, name)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/contrib/lite/python/interpreter_wrapper/tensorflow_wrap_interpreter_wrapper.py", line 74, in _swig_getattr
    return _swig_getattr_nondynamic(self, class_type, name, 0)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/contrib/lite/python/interpreter_wrapper/tensorflow_wrap_interpreter_wrapper.py", line 69, in _swig_getattr_nondynamic
    return object.__getattr__(self, name)
AttributeError: type object 'object' has no attribute '__getattr__'

freedomtan commented:
@jdinkel88 maybe you forgot to install the newly built pip wheel (remember to `pip3 install -U YOUR_NEWLY_BUILT_PIP_WHEEL`) or ran a wrong python interpreter (make sure that the python interpreter uses the newly installed one)?

dpsm commented:
@petewarden can you review this? RPI right now runs the apps using the python binding on single core :(

dpsm commented:
👍

freedomtan commented:
@miaout17 can you help review this? it seems some guys using RPI really want this.

dpsm commented:
@miaout17 is there anyone else you can nominate as reviewer in case you don't have cycles?

dpsm reviewed (dismissed):
Verified on rpi that it indeed uses more than a single core

jdduke commented on tensorflow/lite/python/interpreter.py:
```diff
@@ -290,3 +290,6 @@ def invoke(self):
 
   def reset_all_variables(self):
     return self._interpreter.ResetVariableTensors()
+
```
This method deserves documentation.

freedomtan replied:
Thanks for the review. Added simple description, rebased, and pushed.

Commit:
add simple description of set_num_threads()

dpsm commented:
@jdduke any more changes required?

jdduke commented:
I defer to @miaout17 for final review, but in general looks good.

dpsm commented:
@jdduke anyone else other than @miaout17 can review this? Running tflite single threaded on CPUs with multiple cores is quite constraining.

jdduke commented on tensorflow/lite/python/interpreter.py:
```diff
@@ -292,3 +292,14 @@ def invoke(self):
+  def set_num_threads(self, i):
+    """Set number of threads used by TFLite kernels.
+
+    If not set, kernels are running single-threaded. Note that currently,
```
"If not set, kernels will run with a default threading configuration, which may be single-threaded"

We actually multi-thread float models by default.

freedomtan replied:
Good to know this. Thanks.

jdduke replied:
Let's remove this API since we have the constructor path.

jdduke reviewed (changes requested):
Looks good, just one nit on the comment.

dpsm commented:
@jdduke any insight into what remains to get this merged? Not sure why CI is failing.

jdduke commented:
It hit a snag in our submit pipeline due to a lint issue. Working on resolving now, sorry for the delay. Should get it in for the upcoming release.

dpsm commented:
@jdduke when is this going to be merged any idea?

jdduke commented:
Looks like it got hung up on one of our import scripts, will try to force it through today. Thanks for your patience.

dpsm commented:
@jdduke build still broken ?

jdduke commented:
We're having some internal discussion about whether this should be a mutable property of the Interpreter or if it should be a constructor arg. If you have a strong opinion either way let us know, we're trying to move away from mutable properties as they complicate the codebase and runtiime (i.e., we're considering deprecating the native SetNumThreads API).

freedomtan committed:
Merge branch 'master' into export_SetNumThreads_to_tflite_python

freedomtan commented:
@googlebot I fixed it.

mellanox-github commented:
Can one of the admins verify this patch?

rthadur commented:
@freedomtan can you please resolve conflicts.

freedomtan committed:
Merge branch 'master' into export_SetNumThreads_to_tflite_python

rthadur commented:
@jdduke can you please review new changes ?

jdduke commented:
Defer to @miaout17 for approval.

jdduke commented on tensorflow/lite/python/interpreter.py:
```diff
@@ -456,6 +456,16 @@ def invoke(self):
+    return self._interpreter.SetNumThreads(i)
 
 class InterpreterWithCustomOps(Interpreter):
   """Interpreter interface for TensorFlow Lite Models that accepts custom ops.
```
What about making the number of threads an argument to the constructor instead? We're trying to move away from making non-trivial properties mutable after construction.

freedomtan replied:
sure, will do it.

freedomtan replied:
Does creating a `Options` argument that contains a `numberOfThreads` option as what you did in Java binding [1] make sense to you? I mean to make it more general instead of number of threads specific.

[1] https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/java/src/main/java/org/tensorflow/lite/Interpreter.java#L233-L235

jdduke replied:
Python has much better handling of named, selective arguments for constructors (and functions generally), so I don't think we need an Options construct here. I think we can just add as an other named argument.

freedomtan replied:
OK. I'll add a `num_threads` argument with default value 1.

freedomtan replied:
rebased, cleaned up a bit, and moved the number of threads setting to the constructor  of `Interpreter`

freedomtan committed:
cleanup

freedomtan committed:
set number of threads in constructor

freedomtan committed:
Merge branch 'master' into export_SetNumThreads_to_tflite_python

jdduke commented on tensorflow/lite/python/interpreter.py:
```diff
@@ -228,6 +232,8 @@ def __init__(self,
         self._interpreter.ModifyGraphWithDelegate(
             delegate._get_native_delegate_pointer())  # pylint: disable=protected-access
 
+    self._interpreter.SetNumThreads(num_threads)
```
Let's move this call before setting the delegates.

jdduke commented on tensorflow/lite/python/interpreter.py:
```diff
@@ -183,7 +183,8 @@ class Interpreter(object):
                model_content=None,
-               experimental_delegates=None):
+               experimental_delegates=None,
+               num_threads=1):
```
Let's keep the default None, so it remains implementation-dependent, and we skip setting the threads if it's None.

freedomtan replied:
done

freedomtan committed:
address review concerns

1. change default value of num_threads to Non
2. set num_threads before delegate
3. check the type of num_threads before setting it

duxiaochao commented:
We test this multi-threading whl on the label_image example and it works, but we test it on the lstm or rnn examples and it doesn't work. Can you tell me why it is? Thanks.

freedomtan commented:
@duxiaochao First, rnn and lstm are fundamentally hard to parallelize. Second, as far as I know, only few kernels in TFLite are multithreaded.

freedomtan committed:
update goldens for CI

we changed Python API, so we have to update goldens accordingly

vonchenplus commented:
Hello guys,
When can this patch be merged? Now testing with Python is too slow.

freedomtan commented:
@rthadur and @gbaned I don't know why there is "cla/google" problem. What can I do?

jdduke commented on tensorflow/lite/python/interpreter.py:
```diff
@@ -192,6 +193,9 @@ def __init__(self,
         [TfLiteDelegate](https://www.tensorflow.org/lite/performance/delegates)
         objects returned by lite.load_delegate().
+      num_threads: Set the number of threads used by TFLite kernels.
+        If not set, kernels are running single-threaded. Note that currently,
```
This should instead say:

"If not set, the number of threads used for inference is implementation-dependent. Note that..."

freedomtan replied:
Thanks, done.

jdduke reviewed (changes requested):
This is held up on internal API review, which is quite strict. There was some feedback on the API docs.

gbaned commented:
@freedomtan Can you please resolve conflicts? Thanks!

freedomtan committed:
Merge branch 'master' into export_SetNumThreads_to_tflite_python

rthadur commented:
@jdduke can you please review new changes ?

Commit:
Add SetNumThreads definition for pybind11

Mehanik commented:
Hi! Thank you for very useful feature. It helps me to increase ssd-mobilenet performance 4 times on Raspberry Pi 4.

I'm tried to build tflite from this branch and to use this feature and it seems that `SetNumThreads` defenition for pybind11 is missed. Here is a patch with which I was able to fix the problem.
[0001-Add-SetNumThreads-defenition-for-pybind11.txt](https://github.com/tensorflow/tensorflow/files/4473197/0001-Add-SetNumThreads-defenition-for-pybind11.txt)

freedomtan commented:
@Mehanik Thanks. Yes, TensorFlow switched from swig to pybind11. I'll rebase and apply your patch.

freedomtan committed:
clang-format and add doc

freedomtan committed:
Merge branch 'master' into export_SetNumThreads_to_tflite_python

freedomtan commented:
@Mehanik it seems you need to sign CLA.

freedomtan commented:
@Mehanik are you going to sign the CLA? Or, can I reformat the patch by removing you from the Author and putting your name in the git commit comment?

Mehanik commented:
@freedomtan sorry for the long wait. I'll try to sign the CLA today

Mehanik commented:
@googlebot I fixed it..

Mehanik commented:
@googlebot I consent.

Mehanik commented:
@googlebot I consent.

teijeong commented on tensorflow/lite/examples/python/label_image.py:
```diff
@@ -19,13 +19,11 @@
-
 import tensorflow as tf # TF2
-
+from PIL import Image
```
Is this needed?

freedomtan replied:
yes, to read an image and resize it.

teijeong replied:
Oh yeah, I was wondering why this one marked as added - and I noticed that this is a more-than-a-year-old PR :( (sorry about that)

teijeong commented on tensorflow/lite/examples/python/label_image.py:
```diff
@@ -57,9 +55,15 @@ def load_labels(filename):
       help='input standard deviation')
+  parser.add_argument(
+      '--num_threads',
+      default=1,
```
How about adding `type=int` here, so that line 66 can be just `num_threads=args.num_threads`?

teijeong commented on tensorflow/lite/examples/python/label_image.py:
```diff
@@ -93,3 +99,5 @@ def load_labels(filename):
     else:
       print('{:08.6f}: {}'.format(float(results[i] / 255.0), labels[i]))
+
+  print("time: ", stop_time - start_time)
```
nit: wrap string with single quote
remove extra space after ":", the print statement adds space between arguments.

Preferably would be nicer to format this as something like `'time: {:.3f}ms'.format((stop_time - start_time) * 1000)`

teijeong commented on tensorflow/lite/python/interpreter.py:
```diff
@@ -516,7 +525,6 @@ def invoke(self):
   def reset_all_variables(self):
     return self._interpreter.ResetVariableTensors()
 
-
```
re-add newline

teijeong commented on tensorflow/lite/python/interpreter_wrapper/interpreter_wrapper.cc:
```diff
@@ -688,6 +688,12 @@ PyObject* InterpreterWrapper::ResetVariableTensors() {
   Py_RETURN_NONE;
 }
 
+PyObject* InterpreterWrapper::SetNumThreads(int i) {
```
give this more descriptive name :)

teijeong commented on tensorflow/lite/python/interpreter_wrapper/interpreter_wrapper.h:
```diff
@@ -87,6 +87,8 @@ class InterpreterWrapper {
   // should be the interpreter object providing the memory.
   PyObject* tensor(PyObject* base_object, int i);
 
+  PyObject* SetNumThreads(int i);
```
ditto

teijeong commented on tensorflow/lite/python/interpreter_wrapper/interpreter_wrapper_pybind11.cc:
```diff
@@ -145,5 +145,13 @@ PYBIND11_MODULE(_pywrap_tensorflow_interpreter_wrapper, m) {
+          )pbdoc")
+      .def(
+          "SetNumThreads",
+          [](InterpreterWrapper& self, int i) {
```
ditto

teijeong commented on tensorflow/lite/python/interpreter.py:
```diff
@@ -218,6 +222,11 @@ def __init__(self,
+    if num_threads:
+      if not isinstance(num_threads, int):
+        raise ValueError('type of num_threads should be int')
+      self._interpreter.SetNumThreads(num_threads)
```
Can you add validation for num_threads here?

freedomtan committed:
Merge branch 'master' into export_SetNumThreads_to_tflite_python

freedomtan committed:
address review commments

freedomtan committed:
more modifications

teijeong commented on tensorflow/lite/examples/python/label_image.py:
```diff
@@ -100,4 +100,5 @@ def load_labels(filename):
       print('{:08.6f}: {}'.format(float(results[i] / 255.0), labels[i]))
 
-  print("time: ", stop_time - start_time)
+  #print("time: ", stop_time - start_time)
```
please remove :)

freedomtan replied:
oops, that's embarrassing. Thanks.

teijeong replied:
No worries, it happens all the time for me too ;)

teijeong reviewed (dismissed):
Looks good, just one nit ;) Thanks!

freedomtan committed:
clean up

jdduke commented on tensorflow/lite/python/interpreter.py:
```diff
@@ -181,6 +182,9 @@ def __init__(self,
         [TfLiteDelegate](https://www.tensorflow.org/lite/performance/delegates)
           objects returned by lite.load_delegate().
+      num_threads: Set the number of threads used by TFLite kernels.
+        If not set, kernels are running single-threaded. Note that currently,
```
This should really be:

"Sets the number of threads used by the interpreter and available to CPU kernels. If not set, the interpreter will use an implementation-dependent default number of threads. Currently, only a subset of kernels, such as conv, support multi-threading."

freedomtan replied:
done

jdduke commented on tensorflow/lite/examples/python/label_image.py:
```diff
@@ -57,9 +55,15 @@ def load_labels(filename):
       help='input standard deviation')
+  parser.add_argument(
+      '--num_threads',
+      default=1, type=int,
```
Would prefer that the default be None, if possible?

freedomtan replied:
done

freedomtan committed:
address review comments

rthadur commented:
@jdduke gentle ping if we need more changes here.

jdduke reviewed (changes requested):
Can you add a test or two in interpreter_test.py? Validate that it throws an error for invalid input, and that 2 threads works as expected?

srjoglekar246 commented:
I can add the tests internally, so that we don't have to reimport the changes. Will do it by tomorrow and get this in.

metanav commented:
How can I make a pip package to test it on RPI 4?

teijeong commented:
@metanav Would you try this? https://github.com/tensorflow/tensorflow/tree/master/tensorflow/lite/tools/pip_package
